### PR TITLE
types: add a interface to eval the memory cost of datum array

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
As the title says.

### What is changed and how it works?
N/A

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

N/A

Related changes

N/A

Release note

N/A

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -2057,3 +2058,30 @@ func ChangeReverseResultByUpperLowerBound(
...
+// or two-dimensional datum array.
+func EstimatedMemUsage(array []Datum, numOfRows int) int64 {
+	if numOfRows == 0 {
+		return 0
+	}
+	var bytesConsumed int
+	for _, d := range array {
+		switch d.Kind() {
+		case KindString, KindBytes, KindBinaryLiteral, KindMysqlJSON, KindMysqlEnum, KindMysqlSet, KindMysqlBit:
+			bytesConsumed += len(d.b)
```
Comment (SunRunAway) [inline types/datum.go]: Suggest to add `len(d.b)` in any cases.

File: types/datum_test.go [hunk 2]
```diff
@@ -396,6 +397,26 @@ func newRetTypeWithFlenDecimal(tp byte, flen int, decimal int) *FieldType {
...
 }
 
+func (ts *testDatumSuite) TestEstimatedMemUsage(c *C) {
+	b := []byte{'a', 'b', 'c', 'd'}
+	enum := Enum{Name: "a", Value: 1}
+	datumArray := []Datum{
+		NewIntDatum(1),
+		NewFloat64Datum(1.0),
+		NewFloat32Datum(1.0),
+
```
Comment (qw4990) [inline types/datum_test.go]: Unnecessary blank line.

File: types/datum_test.go [hunk 3]
```diff
@@ -396,6 +397,26 @@ func newRetTypeWithFlenDecimal(tp byte, flen int, decimal int) *FieldType {
 	}
 }
 
+func (ts *testDatumSuite) TestEstimatedMemUsage(c *C) {
+	b := []byte{'a', 'b', 'c', 'd'}
+	enum := Enum{Name: "a", Value: 1}
+	datumArray := []Datum{
```
Comment (qw4990) [inline types/datum_test.go]: It's better to add a new `Enum` whose underlying type is `Time` into it.

Comment (XuHuaiyu) [reply types/datum_test.go]: Refer [MySQL manual](https://dev.mysql.com/doc/refman/8.0/en/enum.html): An ENUM is a **string** object with a value chosen from a list of permitted values

**Pull request merged**